### PR TITLE
Fix ssd cache full related bugs

### PR DESCRIPTION
**Pull request by MahatiC**
- Move code related to `m_free_lanes` and `m_free_log_entries` to rwl
- Remove unnecessary `m_log_entries` check in `process_work` function

Signed-off-by: Mahati Chamarthy <mahati.chamarthy@intel.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by MahatiC** [comment]
This PR still requires some code cleanup, hence WIP tag.

File: src/librbd/cache/pwl/ssd/WriteLog.cc
Comment (idryomov) [inline src/librbd/cache/pwl/ssd/WriteLog.cc:458]: Which uninitialized variable are you referring to the initial comment?

I wanted to patch it for my local tests so that there is one less bug to run into and thought it was this one, but it is initialized at the top of `append_ops()` and you just moved that initialization here.

Comment (MahatiC) [reply src/librbd/cache/pwl/ssd/WriteLog.cc:458]: You maybe right, I must've missed that I removed the variable in `append_ops` and added it back up here when I encountered a bug.

I've marked and commented on a couple of variables that I had changed, incase you want to patch this change.

File: src/librbd/cache/pwl/ssd/WriteLog.cc
Comment (MahatiC) [inline src/librbd/cache/pwl/ssd/WriteLog.cc:559]: This fixes the bug where `retire_entries` is not triggered and it hangs when ssd cache is full. This is because the code was wrongly making a retire decision based on `m_log_entries`

File: src/librbd/cache/pwl/ssd/WriteLog.cc
Comment (MahatiC) [inline src/librbd/cache/pwl/ssd/WriteLog.cc:618]: `m_log_pool_ring_buffer_size` is correctly updated here and throughout the code

File: src/librbd/cache/pwl/AbstractWriteLog.cc
Comment (MahatiC) [inline src/librbd/cache/pwl/AbstractWriteLog.cc:1450]: This is commented out for now, since this is `rwl` specific checks and will be moved to `rwl` in the revised PR

Comment (idryomov) [reply src/librbd/cache/pwl/AbstractWriteLog.cc:1450]: This actually helps a lot because I was trying to make sense of `m_free_log_entries` logic for the `ssd` PR I'm working on and couldn't...  With `m_free_log_entries` out of the way I should be able to post on Monday.

**Comment by MahatiC** [comment]
@idryomov I quickly pushed an unclean PR, just so that you could take a look at the other changes that maybe impacting your scenarios. Please hold off patching this PR for testing yet. I'll clean this up properly and push a revised PR.

**Comment by MahatiC** [comment]
This is a cleaned up PR and will pass the unit tests. Haven't run the performance/cache full related tests yet.

File: src/librbd/cache/pwl/ssd/WriteLog.cc
Comment (MahatiC) [reply src/librbd/cache/pwl/ssd/WriteLog.cc:546]: Actually `m_log_entries` size is irrelevant for SSD cache, therefore this check is is not meaningful and not necessary.

**Comment by MahatiC** [comment]
@idryomov I've cleaned this PR and I think this is ready to be merged. Could you review please?

File: src/librbd/cache/pwl/ssd/WriteLog.cc
Comment (idryomov) [inline src/librbd/cache/pwl/ssd/WriteLog.cc:295]: Did you mean to bump these counters unconditionally, even if `alloc_succeeds` is false?  This messes up the cache pretty quickly and I/O just hangs.

**Review by idryomov — CHANGES_REQUESTED** [review]
Hi Mahati,

I've already gotten rid of log entry counting in ssd mode in a less invasive way:

https://paste.centos.org/view/raw/83657059

Previous versions of this PR were causing crashes because of bad `m_log_pool_config_size` modulo arithmetic changes.  You seem to have dropped those, but this version still appears to be broken to me.

**Comment by idryomov** [comment]
Was moving `m_free_lanes` and `m_free_log_entries` logic the only full handling fix or did you have something else in mind?

File: src/librbd/cache/pwl/ssd/WriteLog.cc
Comment (MahatiC) [reply src/librbd/cache/pwl/ssd/WriteLog.cc:295]: Ugh, I'm sorry. No, I didn't intend to remove the condition of `alloc_succeeds`. I'll fix it.

Comment (idryomov) [reply src/librbd/cache/pwl/ssd/WriteLog.cc:295]: It's fixed (or rather avoided altogether) in my patch.  Did you see the paste?

I prefer that approach because it minimizes the risk of regression for rwl mode -- only ssd mode is patched and there is a lot less code churn.

**Comment by MahatiC** [comment]
> Was moving `m_free_lanes` and `m_free_log_entries` logic the only full handling fix or did you have something else in mind?

Cleaning up and moving the rwl code was what I wanted to do with this PR. Would you like me to fix this PR wrt to your review? Or do you already have a PR that you'd like to push?

**Comment by idryomov** [comment]
I will include the pasted patch in my next PR and CC you on it.  I was asking to make sure that I'm not missing something else in this PR.

File: src/librbd/cache/pwl/ssd/WriteLog.cc
Comment (MahatiC) [reply src/librbd/cache/pwl/ssd/WriteLog.cc:295]: I did see the paste. You're right that there's less code churn but I was wondering if it would be more clear to remove them altogether in `ssd` code as that would make it more clear that they do not belong in `ssd` logic - I think this would help future code reviewers/maintainers?

Comment (idryomov) [reply src/librbd/cache/pwl/ssd/WriteLog.cc:295]: Well, your patch left those fields in `AbstractWriteLog`, so they weren't removed altogether -- `ssd/WriteLog` could still access them.  Doing it properly would cause even more code churn and I'd like to avoid that until everything is stable.

Comment (MahatiC) [reply src/librbd/cache/pwl/ssd/WriteLog.cc:295]: > Well, your patch left those fields in `AbstractWriteLog`, so they weren't removed altogether -- `ssd/WriteLog` could still access them. 

Right. I could correct those, not saying this PR should be merged as is. It's a thought I had when I looked at the paste you shared
>Doing it properly would cause even more code churn and I'd like to avoid that until everything is stable.

okay

**Comment by idryomov** [comment]
Since there weren't any other fixes in this PR, let's close?

**Comment by MahatiC** [comment]
closing as these changes are addressed in the https://github.com/ceph/ceph/pull/41354

**Pull request closed without merging**
